pkg/rancher-desktop/assets/scripts/nginx.conf

```diff
@@ -0,0 +1,175 @@
+worker_processes  auto;
+
+events {
+    worker_connections  1024;
+}
+
+http {
+    map_hash_bucket_size 128;
+    include       mime.types;
+    default_type  application/octet-stream;
+
+    log_format proxy escape=json
+      '{'
+        '"access_time":"$time_local",'
+        '"request":"$request",'
+        '"status":"$status",'
+        '"bytes_sent":"$body_bytes_sent",'
+        '"host":"$host",'
+        '"ssl_protocol":"$ssl_protocol",'
+        '"connect_host":"$connect_host",'
+        '"connect_port":"$connect_port",'
+      '}';
+
+    log_format mitm escape=json
+      '{'
+        '"access_time":"$time_local",'
+        '"method":"$request_method",'
+        '"uri":"$uri",'
+        '"status":"$status",'
+        '"bytes_sent":"$body_bytes_sent",'
+        '"upstream_response_time":"$upstream_response_time",'
+        '"host":"$host",'
+        '"http_host":"$http_host",'
+        '"upstream":"$upstream_addr"'
+      '}';
+
+    gzip off;
+
+    server {
+        listen 3128;
+        listen [::]:3128;
+        server_name proxy;
+
+        access_log logs/proxy.log proxy;
+
+        proxy_connect;
+        proxy_connect_allow all;
+        proxy_connect_address 127.0.0.1:3129;
+        proxy_max_temp_file_size 0;
+
+        # response non-CONNECT requests
+        location / {
+            add_header "Content-type" "text/plain" always;
+            return 200 "Yo docker, wazzup? Let's watch the game!\n";
+        }
+    }
+
+    map "$http_host$uri" $forbidden {
+        default 1;
+        include image-allow-list.conf;
+    }
+
+    server {
+        listen 3129 ssl default_server;
+        server_name mitm;
+
+        access_log logs/access.log mitm;
+
+        # nginx complains if these are not set; we'll clear them again right after
+        ssl_certificate /run/mkcert/localhost.pem;
+        ssl_certificate_key /run/mkcert/localhost-key.pem;
+
+        ssl_certificate_by_lua_block {
+            local ssl = require "ngx.ssl"
+            local name = ssl.server_name()
+
+            local ok, err = ssl.clear_certs()
+            if not ok then
+                ngx.log(ngx.ERR, "failed to clear existing (fallback) certificates")
+                return ngx.exit(ngx.ERROR)
+            end
+
+            local certs_dir = "/run/mkcert/"
+            local cert_file = certs_dir .. name .. ".pem"
+            local key_file = certs_dir .. name .. "-key.pem"
+
+            local my_load_certificate_chain = function()
+                local f = io.open(cert_file, "rb")
+                if f == nil then
+                    local ngx_pipe = require "ngx.pipe"
+                    local cmd = { "/usr/bin/mkcert", "-cert-file", cert_file, "-key-file", key_file, name }
```

Comment:
We're creating one cert file per host name, so if the user is accessing generated host names (e.g. things that would normally be covered by wildcard certs) this could lead to an unintentional DoS. Not sure how much we should care though? Would this cause an issue with epinio?

Annoyingly, wildcard certs can only be single-level…

Reply:
I don't understand the concern, or how wildcards come into play here. What is the DoS scenario?

The proxy is only used by containerd/dockerd/nerdctl accessing container registries. So we should get one cert for each registry the user is pulling images from, which should be a reasonably small number. And we start fresh with an empty certs directory on each boot.

Reply:
A malicious user could (attempt to) pull images from many domains, filling up disk. That's not likely to be a real concern, though; the actual user controls the whole thing. So the only worry would be a malicious image with docker/containerd API access, at which point we're hosed anyway.
So I think we don't need to worry about this.